operate/cli.py

```diff
@@ -663,6 +666,98 @@ async def _stop_service_onchain(request: Request) -> JSONResponse:
             )
         )
 
+    @app.post("/api/services/{service}/onchain/withdraw")
+    @with_retries
+    async def _withdraw_onchain(request: Request) -> JSONResponse:
+        """Withdraw all the funds from a service."""
+        if not operate.service_manager().exists(service=request.path_params["service"]):
+            return service_not_found_error(service=request.path_params["service"])
+        if operate.password is None:
+            return USER_NOT_LOGGED_IN_ERROR
+
+        withdrawal_address = (await request.json()).get("withdrawal_address")
+        if withdrawal_address is None:
+            return JSONResponse(
+                content={"error": "withdrawal_address is required"},
+                status_code=400,
+            )
+
+        try:
+            service_manager = operate.service_manager()
+            service = service_manager.load_or_create(
+                hash=request.path_params["service"]
+            )
+
+            service_manager.terminate_service_on_chain_from_safe(
+                hash=request.path_params["service"],
+                chain_id=service.home_chain_id,
+                is_withdrawing=True,
+            )
+
+            service_manager.drain_service(
+                hash=request.path_params["service"],
+                withdrawal_address=withdrawal_address,
+            )
+
+            # drain OLAS from the master safe
+            chain_config = service.chain_configs[service.home_chain_id]
+            ledger_config = chain_config.ledger_config
+            master_wallet = service_manager.wallet_manager.load(
+                ledger_type=ledger_config.type
+            )
+            ledger_api = master_wallet.ledger_api(
+                chain_type=ledger_config.chain, rpc=ledger_config.rpc
+            )
+            token_instance = registry_contracts.erc20.get_instance(
+                ledger_api=ledger_api,
+                contract_address=OLAS[ledger_config.chain],
+            )
+            balance = token_instance.functions.balanceOf(master_wallet.safe).call()
+            if balance == 0:
+                logger.info(f"No OLAS to drain from master safe: {master_wallet.safe}")
+            else:
+                logger.info(
+                    f"Draining {balance} OLAS out of master safe: {master_wallet.safe}"
+                )
+                transfer_erc20_from_safe(
+                    ledger_api=ledger_api,
+                    crypto=master_wallet.crypto,
+                    safe=t.cast(str, master_wallet.safe),
+                    token=OLAS[ledger_config.chain],
+                    to=withdrawal_address,
+                    amount=balance,
+                )
```

Comment:
Refactor in another PR: This should be done through the wallet/wallet manager, not here on the cli. Consider this for refactoring in another PR, or in the multi-agent branch.